File: haskell/glue/src/Glue/Lib/Bool/Le.hs
module Glue.Lib.Bool.Le where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

le :: IR Eval
le = NativeFunc leImpl

leImpl :: IR Eval -> Eval (IR Eval)
leImpl a = pure $ NativeFunc (leRight a)

leRight :: IR Eval -> IR Eval -> Eval (IR Eval)
leRight a b = case (a, b) of
    (Integer na, Integer nb) -> pure . Bool $ na <= nb
    (Float na, Float nb) -> pure . Bool $ na <= nb
    (Integer na, Float nb) -> pure . Bool $ fromIntegral na <= nb
    (Float na, Integer nb) -> pure . Bool $ na <= fromIntegral nb
    _ -> throwError $ wrongArgumentType ["number", "number"]

File: dart/glue/lib/src/lib/bool/le.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Less than or equal comparison function
/// Mirrors Haskell Glue.Lib.Bool.Le.le exactly
final Ir le = IrNativeFunc(leImpl);

/// Less than or equal comparison implementation
/// Mirrors Haskell Glue.Lib.Bool.Le.leImpl exactly
Eval<Ir> leImpl(Ir a) {
  return Eval.pure(IrNativeFunc(leRight(a)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Bool.Le.leRight exactly
Eval<Ir> Function(Ir) leRight(Ir a) {
  return (Ir b) => switch ((a, b)) {
    (IrInteger(value: final na), IrInteger(value: final nb)) => Eval.pure(
      IrBool(na <= nb),
    ),
    (IrFloat(value: final na), IrFloat(value: final nb)) => Eval.pure(
      IrBool(na <= nb),
    ),
    (IrInteger(value: final na), IrFloat(value: final nb)) => Eval.pure(
      IrBool(na <= nb),
    ),
    (IrFloat(value: final na), IrInteger(value: final nb)) => Eval.pure(
      IrBool(na <= nb),
    ),
    _ => throwError(wrongArgumentType(['number', 'number'])),
  };
}
